standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java

```diff
@@ -1424,30 +1422,36 @@ public void visit(LeafNode node) throws MetaException {
         return;
       }
 
+      String nodeValue0 = "?";
       // if Filter.g does date parsing for quoted strings, we'd need to verify there's no
       // type mismatch when string col is filtered by a string that looks like date.
-      if (colType == FilterType.Date && valType == FilterType.String) {
-        // Filter.g cannot parse a quoted date; try to parse date here too.
+      if (colType == FilterType.Date) {
         try {
-          nodeValue = MetaStoreUtils.convertStringToDate((String)nodeValue);
+          // check the nodeValue is a valid date
+          nodeValue = MetaStoreUtils.convertDateToString(
```

Comment:
missed that part, why don't we do similar to PartFilterVisitor.java
````
MetaStoreUtils.convertStringToDate((String) nodeValue);
````
and use nodeValue if no exceptions;
if that does some date with TZ string reformatting, can we create a helper method in MetaStoreUtils

Reply:
if we return the `Date` value from `PartFilterVisitor`, we couldn't convert the date back to the original string in JDO path(we compare the `PART_NAME` on JDO path, which is timezone agnostic), this could result to query difference   as we met in this jira.

Reply:
PartFilterVisitor returns String date representation after validation, isn't it?

Reply:
This is for converting the string date to a timezone based string date, we need it on the direct sql:
https://github.com/apache/hive/blob/master/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/DatabaseProduct.java#L263-L269